File: tempestweb/native/orientation.py
"""Native screen-orientation capability over the Screen Orientation Web API.

:func:`lock` / :func:`unlock` send the ``orientation.lock`` / ``orientation.unlock``
``native_call`` envelopes and :func:`state` reads the current orientation;
``client/native/orientation.js`` calls ``screen.orientation.lock`` /
``screen.orientation.unlock`` and reads ``screen.orientation.{type,angle}``. Locking
requires fullscreen on most browsers.
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from dataclasses import dataclass

from tempestweb.native.dispatch import native_events, send_native_call

__all__ = ["OrientationState", "lock", "state", "unlock", "watch"]


@dataclass(frozen=True)
class OrientationState:
    """The current screen orientation.

    Attributes:
        type: The orientation type (e.g. ``"portrait-primary"``,
            ``"landscape-primary"``).
        angle: The orientation angle in degrees (``0``, ``90``, ``180``, ``270``).
    """

    type: str
    angle: int


async def lock(kind: str) -> bool:
    """Lock the screen to a given orientation.

    Args:
        kind: The orientation to lock to (e.g. ``"portrait"``, ``"landscape"``,
            ``"portrait-primary"``).

    Returns:
        ``True`` if the lock was applied.

    Raises:
        NativeError: If the lock is refused (``permission_denied``) or unsupported
            (``unavailable``) — e.g. not in fullscreen.
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("orientation.lock", {"kind": kind})
    return bool(value.get("locked", False))


async def unlock() -> None:
    """Release any screen-orientation lock.

    Raises:
        NativeError: If the API is unavailable (``unavailable``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    await send_native_call("orientation.unlock", {})


async def state() -> OrientationState:
    """Report the current screen orientation.

    Returns:
        The current :class:`OrientationState`.

    Raises:
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("orientation.state", {})
    return OrientationState(
        type=str(value.get("type", "")),
        angle=int(value.get("angle", 0)),
    )


async def watch() -> AsyncIterator[OrientationState]:
    """Stream screen-orientation changes from the browser (event channel / T-EV).

    Yields a fresh :class:`OrientationState` whenever the device is rotated, until
    the ``async for`` loop is exited (which closes the subscription). Consume it
    with::

        async for orient in native.orientation.watch():
            app.set_state(lambda s: setattr(s, "orientation", orient))

    Yields:
        Each updated :class:`OrientationState`.

    Raises:
        NativeError: If the browser reports the subscription failed.
        BrowserUnavailableError: If no bridge is installed, or the installed bridge
            does not support the event channel.
    """
    async for value in native_events("orientation.watch", {}):
        yield OrientationState(
            type=str(value.get("type", "")),
            angle=int(value.get("angle", 0)),
        )

File: client/native/orientation.js
// native/orientation.js — Screen Orientation API glue.
//
// `screen.orientation.lock(kind)` requires fullscreen on most platforms and is
// unsupported on desktop; `unlock` and `state` are best-effort reads. Reached via
// `deps.screen` so the router stays testable under jsdom.

import { CapabilityError } from "./index.js";

/**
 * Lock the screen orientation to a kind (e.g. "portrait", "landscape").
 * @param {{kind:string}} args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{locked:boolean}>}
 * @throws {CapabilityError} unavailable / not_allowed.
 */
export async function orientationLock(args, deps) {
  const screen = deps.screen || /** @type {any} */ (globalThis).screen;
  if (!screen || !screen.orientation || typeof screen.orientation.lock !== "function") {
    throw new CapabilityError("unavailable", "the Screen Orientation lock API is not available");
  }
  try {
    await screen.orientation.lock(args.kind);
  } catch (err) {
    throw new CapabilityError("not_allowed", err && err.message);
  }
  return { locked: true };
}

/**
 * Release any orientation lock.
 * @param {Object} _args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<Object>}
 * @throws {CapabilityError} unavailable.
 */
export async function orientationUnlock(_args, deps) {
  const screen = deps.screen || /** @type {any} */ (globalThis).screen;
  if (!screen || !screen.orientation || typeof screen.orientation.unlock !== "function") {
    throw new CapabilityError("unavailable", "the Screen Orientation API is not available");
  }
  screen.orientation.unlock();
  return {};
}

/**
 * Report the current orientation type and angle.
 * @param {Object} _args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{type:string, angle:number}>}
 * @throws {CapabilityError} unavailable.
 */
export async function orientationState(_args, deps) {
  const screen = deps.screen || /** @type {any} */ (globalThis).screen;
  if (!screen || !screen.orientation) {
    throw new CapabilityError("unavailable", "the Screen Orientation API is not available");
  }
  return {
    type: screen.orientation.type,
    angle: screen.orientation.angle,
  };
}
